### PR TITLE
Making event-tracker as a k8s controller to run by custom resource(Policy)

### PR DESCRIPTION
**Pull request by rajdas98**
## Proposed changes

Changes:
- Making event-tracker as a k8s controller to run by a custom resource called eventtrackerpolicy:
```yaml
apiVersion: eventtracker.litmuschaos.io/v1
kind: EventTrackerPolicy
metadata:
  name: eventtrackerpolicy-sample
spec:
  # Add fields here
  condition_type: "or"
  conditions:
  - key: "spec.replicas"
    value: "1"
    operator: EqualTo
  - key: "spec.template.spec.containers[0].image"
    value: "nginx:1.18"
    operator: EqualTo
```
- Supported operators
  - EqualTo
  - NotEqualTo
  - LessThan
  - GreaterThan
  - LessThanEqualTo
  - GreaterThanEqualTo

## Types of changes

What types of changes does your code introduce to Litmus? Put an `x` in the boxes that apply
- [x] New feature (non-breaking change which adds functionality)
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation Update (if none of the other choices applies)

## Checklist

Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code.
- [ ] I have read the [CONTRIBUTING](https://github.com/litmuschaos/litmus/blob/master/CONTRIBUTING.md) doc
- [ ] I have signed the commit for DCO to be passed.
- [ ] Lint and unit tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works (if appropriate)
- [ ] I have added necessary documentation (if appropriate)

## Dependency
- Please add the links to the dependent PR need to be merged before this (if any).

### CODE REVIEW

File: litmus-portal/cluster-agents/event-tracker/PROJECT [hunk 1]
```diff
@@ -0,0 +1,7 @@
+domain: litmuschaos.io
+repo: github.com/litmuschaos/litmus/litmus-portal/cluster-agents/event-tracker
+resources:
+- group: eventtracker
+  kind: EventTrackerPolicy
+  version: v1
+version: "2"
```
Comment (ishangupta-ds) [inline litmus-portal/cluster-agents/event-tracker/PROJECT:7]: can it be `version: "1"` @rajdas98

File: litmus-portal/cluster-agents/event-tracker/controllers/eventtrackerpolicy_controller.go [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+	"log"
+	"strings"
+	"sync"
+
+	"github.com/go-logr/logr"
+	eventtrackerv1 "github.com/litmuschaos/litmus/litmus-portal/cluster-agents/event-tracker/api/v1"
+	"github.com/litmuschaos/litmus/litmus-portal/cluster-agents/event-tracker/pkg/utils"
+	"k8s.io/apimachinery/pkg/runtime"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
```
Comment (ishangupta-ds) [inline litmus-portal/cluster-agents/event-tracker/controllers/eventtrackerpolicy_controller.go:31]: are these imports sorted?

File: litmus-portal/cluster-agents/event-tracker/PROJECT (hunk 1 above)
Comment (rajdas98) [reply litmus-portal/cluster-agents/event-tracker/PROJECT:7]: it is auto generated

File: litmus-portal/cluster-agents/event-tracker/controllers/eventtrackerpolicy_controller.go (hunk 2 above)
Comment (rajdas98) [reply litmus-portal/cluster-agents/event-tracker/controllers/eventtrackerpolicy_controller.go:31]: not sure, the ci check is also passed.

**Review by ishangupta-ds — APPROVED** [review]

**Review by ishangupta-ds — APPROVED** [review]
Changes look good but constants can be added instead of strings for known configurations in controller and the informer, like litmus annotations and `conditionpassed` etc. @rajdas98

File: litmus-portal/cluster-agents/event-tracker/pkg/utils/utils.go [hunk 3]
```diff
@@ -0,0 +1,246 @@
...
+			mar, err := json.Marshal(sts)
+			if err != nil {
+				return err
+			}
+
+			err = json.Unmarshal(mar, &dataInterface)
+		} else {
+			return errors.New("Resource not supported")
+		}
+
```
Comment (ishangupta-ds) [inline litmus-portal/cluster-agents/event-tracker/pkg/utils/utils.go:155]: you may use switch case here so the string operation is performed only once

File: litmus-portal/cluster-agents/event-tracker/pkg/utils/utils.go [hunk 4]
```diff
@@ -0,0 +1,246 @@
...
+	} else if operator == "LessThan" {
+		return key < value
+	} else if operator == "GreaterThan" {
+		return key > value
+	} else if operator == "GreaterThanEqualTo" {
+		return key >= value
+	} else if operator == "LessThanEqualTo" {
+		return key <= value
+	}
+	return false
```
Comment (ishangupta-ds) [inline litmus-portal/cluster-agents/event-tracker/pkg/utils/utils.go:50]: here too a switch case would look cleaner

File: litmus-portal/cluster-agents/event-tracker/pkg/utils/informers.go [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+				if err != nil {
+					log.Print(err)
+					return
+				}
+
+				if depNewObj.GetResourceVersion() != depOldObj.GetResourceVersion() && !reflect.DeepEqual(newDep, oldDep) {
+					var worflowid = depNewObj.GetAnnotations()["litmuschaos.io/workflow"]
+					if depNewObj.GetAnnotations()["litmuschaos.io/gitops"] == "true" && worflowid != "" {
+						log.Print("EventType: Update")
+						log.Printf("GitOps Notification for workflowID: %s, ResourceType: %s, ResourceName: %s, ResourceNamespace: %s", worflowid, "Deployment", depNewObj.Name, depNewObj.Namespace)
```
Comment (ishangupta-ds) [inline litmus-portal/cluster-agents/event-tracker/pkg/utils/informers.go:53]: can we combine them and use a `\n` if possible

File: litmus-portal/cluster-agents/event-tracker/pkg/utils/informers.go [hunk 6]
```diff
@@ -0,0 +1,175 @@
...
+				if err != nil {
+					log.Print(err)
+					return
+				}
+
+				if stsNewObj.GetResourceVersion() != stsOldObj.GetResourceVersion() && !reflect.DeepEqual(newSts, oldSts) {
+					var worflowid = stsNewObj.GetAnnotations()["litmuschaos.io/workflow"]
+					if stsNewObj.GetAnnotations()["litmuschaos.io/gitops"] == "true" && worflowid != "" {
+						log.Print("EventType: Update")
+						log.Printf("GitOps Notification for workflowID: %s, ResourceType: %s, ResourceName: %s, ResourceNamespace: %s", worflowid, "StateFulSet", stsNewObj.Name, stsNewObj.Namespace)
```
Comment (ishangupta-ds) [inline litmus-portal/cluster-agents/event-tracker/pkg/utils/informers.go:109]: same comment as above

File: litmus-portal/cluster-agents/event-tracker/pkg/utils/informers.go [hunk 7]
```diff
@@ -0,0 +1,175 @@
...
+				if err != nil {
+					log.Print(err)
+					return
+				}
+
+				if dsNewObj.GetResourceVersion() != dsOldObj.GetResourceVersion() && !reflect.DeepEqual(newDm, oldDm) {
+					var worflowid = dsNewObj.GetAnnotations()["litmuschaos.io/workflow"]
+					if dsNewObj.GetAnnotations()["litmuschaos.io/gitops"] == "true" && worflowid != "" {
+						log.Print("EventType: Update")
+						log.Printf("GitOps Notification for workflowID: %s, ResourceType: %s, ResourceName: %s, ResourceNamespace: %s", worflowid, "DaemonSet", dsNewObj.Name, dsNewObj.Namespace)
```
Comment (ishangupta-ds) [inline litmus-portal/cluster-agents/event-tracker/pkg/utils/informers.go:162]: same here

**Review by ishangupta-ds — COMMENTED** [review]
@rajdas98 PTAL

File: litmus-portal/cluster-agents/event-tracker/pkg/utils/utils.go (hunk 3 above)
Comment (rajdas98) [reply litmus-portal/cluster-agents/event-tracker/pkg/utils/utils.go]: Done

File: litmus-portal/cluster-agents/event-tracker/pkg/utils/utils.go (hunk 4 above)
Comment (rajdas98) [reply litmus-portal/cluster-agents/event-tracker/pkg/utils/utils.go]: Done

File: litmus-portal/cluster-agents/event-tracker/pkg/utils/informers.go (hunk 7 above)
Comment (rajdas98) [reply litmus-portal/cluster-agents/event-tracker/pkg/utils/informers.go]: Done

**Comment by rajdas98** [comment]
> Changes look good but constants can be added instead of strings for known configurations in controller and the informer, like litmus annotations and `conditionpassed` etc. @rajdas98

Added a constant string ConditionPassed

File: litmus-portal/cluster-agents/event-tracker/pkg/utils/utils.go [hunk 8]
```diff
@@ -0,0 +1,250 @@
...
+
+			err = json.Unmarshal(mar, &dataInterface)
+		default:
+			return errors.New("Resource not supported")
+		}
+
+		check := conditionChecker(etp, dataInterface)
+		var result string
+		if check == true {
+			result = "ConditionPassed"
```
Comment (ishangupta-ds) [inline litmus-portal/cluster-agents/event-tracker/pkg/utils/utils.go:163]: please use the constant already defined @rajdas98

File: litmus-portal/cluster-agents/event-tracker/pkg/utils/utils.go [hunk 9]
```diff
@@ -0,0 +1,250 @@
...
+		} else if check == false {
+			result = "ConditionFailed"
```
Comment (ishangupta-ds) [inline litmus-portal/cluster-agents/event-tracker/pkg/utils/utils.go:165]: please add a constant for this one too

**Review by ishangupta-ds — CHANGES_REQUESTED** [review]

**Review by ishangupta-ds — APPROVED** [review]
LGTM

**Pull request merged**